File: examples/python/test_preprocessing.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import numpy as np
import time
from random import random
import open3d as o3d
import cv2
import copy

import misc3d as m3d
from utils import np2o3d
from IPython import embed

vis = o3d.visualization.Visualizer()
vis.create_window("Preprocessing", 1920, 1200)

color_img = cv2.imread(
    '/home/yuecideng/WorkSpace/Sources/Misc3D/scripts/output/color/color_0.png'
)
color_img = cv2.cvtColor(color_img, cv2.COLOR_BGR2RGB)

depth_img = cv2.imread(
    '/home/yuecideng/WorkSpace/Sources/Misc3D/scripts/output/depth/depth_0.png',
    cv2.IMREAD_ANYDEPTH)

depth = o3d.geometry.Image(depth_img)
color = o3d.geometry.Image(color_img)

rgbd = o3d.geometry.RGBDImage.create_from_color_and_depth(
    color, depth, convert_rgb_to_intensity=False)

pinhole_camera_intrinsic = o3d.camera.PinholeCameraIntrinsic(
    848, 480, 598.7568, 598.7568, 430.3443, 250.244)
pcd = o3d.geometry.PointCloud.create_from_rgbd_image(
    rgbd, pinhole_camera_intrinsic, project_valid_depth_only=False)

t1 = time.time()
# pcd = m3d.preprocessing.crop_roi_pointcloud(pcd, (200, 50, 600, 400), (848, 480))
pcd = m3d.preprocessing.project_into_plane(pcd)
print('time cost: {}'.format(time.time() - t1))

# cv2.rectangle(color_img, (200, 50), (600, 400), (0, 255, 0), 3)
# cv2.imshow('color', color_img)
# cv2.waitKey(1000)
m3d.vis.draw_point_cloud(vis, pcd)
vis.run()

File: examples/cpp/test_preprocessing.cpp
#include <iostream>
#include <memory>

#include <misc3d/preprocessing/filter.h>
#include <misc3d/utils.h>
#include <misc3d/vis/vis_utils.h>
#include <open3d/camera/PinholeCameraIntrinsic.h>
#include <open3d/geometry/Image.h>
#include <open3d/geometry/PointCloud.h>
#include <open3d/geometry/RGBDImage.h>
#include <open3d/io/ImageIO.h>

int main(int argc, char *argv[]) {
    bool ret;
    open3d::geometry::Image depth, color;
    ret = open3d::io::ReadImage("../examples/data/indoor/depth/depth_0.png",
                                depth);
    ret = open3d::io::ReadImage("../examples/data/indoor/color/color_0.png",
                                color);

    auto rgbd = open3d::geometry::RGBDImage::CreateFromColorAndDepth(
        color, depth, 1000, 3.0, false);

    open3d::camera::PinholeCameraIntrinsic intrinsic(
        848, 480, 598.7568, 598.7568, 430.3443, 250.244);

    auto pcd = open3d::geometry::PointCloud::CreateFromRGBDImage(
        *rgbd, intrinsic, Eigen::Matrix4d::Identity(), false);

    misc3d::Timer timer;
    timer.Start();
    auto pcd_roi = misc3d::preprocessing::CropROIPointCloud(
        *pcd, {500, 300, 600, 400}, {848, 480});
    auto pcd_plane = misc3d::preprocessing::ProjectIntoPlane(*pcd);
    std::cout << "Time cost: " << timer.Stop() << std::endl;

    auto vis = std::make_shared<open3d::visualization::Visualizer>();
    vis->CreateVisualizerWindow("Crop ROI", 1920, 1200);
    misc3d::vis::DrawPointCloud(vis, *pcd);
    misc3d::vis::DrawPointCloud(vis, *pcd_roi, {1, 0, 0});
    vis->Run();

    vis = std::make_shared<open3d::visualization::Visualizer>();
    vis->CreateVisualizerWindow("Project into plane", 1920, 1200);
    misc3d::vis::DrawPointCloud(vis, *pcd_plane);
    vis->Run();

    return 0;
}